iceoryx_hoofs/platform/win/source/handle_translator.cpp

```diff
@@ -16,33 +16,78 @@
 
 #include "iceoryx_hoofs/platform/handle_translator.hpp"
 
+#include <iostream>
+
+constexpr int HandleTranslator::INVALID_LINUX_FD;
+
 HandleTranslator& HandleTranslator::getInstance() noexcept
 {
     static HandleTranslator globalHandleTranslator;
     return globalHandleTranslator;
 }
 
-HANDLE HandleTranslator::get(const int handle) const noexcept
+HANDLE HandleTranslator::get(const int linuxFd) const noexcept
 {
-    return m_handleList[static_cast<size_t>(handle)];
+    if (linuxFd == INVALID_LINUX_FD)
+    {
+        return INVALID_HANDLE_VALUE;
+    }
+
+    std::lock_guard<std::mutex> lock(m_mtx);
+
+    auto iter = m_linuxToWindows.find(linuxFd);
+    if (iter == m_linuxToWindows.end())
+    {
+        std::cerr << "Cannot acquire windows file handle for not registered linux file descriptor " << linuxFd
+                  << std::endl;
+        return INVALID_HANDLE_VALUE;
+    }
+
+    return iter->second;
 }
 
-int HandleTranslator::add(HANDLE handle) noexcept
+int HandleTranslator::add(HANDLE windowsHandle) noexcept
 {
-    for (int64_t limit = m_handleList.size(), k = 0; k < limit; ++k)
+    if (windowsHandle == INVALID_HANDLE_VALUE)
+    {
+        return INVALID_LINUX_FD;
+    }
+
+    std::lock_guard<std::mutex> lock(m_mtx);
+
+    int linuxFd = INVALID_LINUX_FD;
+    if (!m_freeFileDescriptors.empty())
+    {
+        linuxFd = m_freeFileDescriptors.front();
+        m_freeFileDescriptors.pop();
+    }
+    else
     {
-        if (m_handleList[k] == nullptr)
-        {
-            m_handleList[k] = handle;
-            return k;
-        }
+        linuxFd = m_currentLinuxFileHandle;
+        ++m_currentLinuxFileHandle;
```

Comment:
Isn't it dangerous to reuse file handles? Wouldn't it be better to e.g. use all file handles up to 2^16 and then do a wraparound?

Reply:
Maybe, but in most platforms we reuse them either way, so when we add extra logic here windows would be the "safer" platform.
I doubt that we gain something then in the long run since we do not have control over the file handle values on other platforms.